### PR TITLE
memberlist: Fix panics when accessing delegate methods before Starting or calling NotifyMsg during service start

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR changes `memberlist.KV` to protect calls to methods from ``memberlist.Delegate` interface. Previously these methods were protected by `initWG` field *during* KV service starting, but not before.

`NotifyMsg` was previously protected by `initWG` field too, but this check was removed in https://github.com/grafana/dskit/pull/110. In this PR, we re-add the check (using `delegateReady`).

Note that under normal conditions clients of `*memberlist.KV` are NOT supposed to call these delegate methods. Only memberlist library calls them via `memberlist.Delegate` interface.

**Which issue(s) this PR fixes**:

Fixes problem reported in https://github.com/grafana/mimir/issues/3732.

**Checklist**
- [x] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by ortuman — APPROVED** [review]
Thanks for the fix Peter! 

Just left a small comment, but in general LGTM. 👍

File: kv/memberlist/memberlist_client.go [hunk 1]
```diff
@@ -233,9 +234,9 @@ type KV struct {
 	provider DNSProvider
 
 	// Protects access to memberlist and broadcasts fields.
-	initWG     sync.WaitGroup
-	memberlist *memberlist.Memberlist
-	broadcasts *memberlist.TransmitLimitedQueue
+	delegateReady atomic.Bool
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply kv/memberlist/memberlist_client.go:237]: Reason was to make methods not crash *before* service is started. Previously we only increased `initWG` in `starting` method, but that also means that before `starting` method was called (by starting a service), `initWG.Wait()` would actually not wait, but proceed immediately -- and crash!

I don't think discarding few initial messages is a big problem:
- normally, only memberlist should call these methods, memberlist should not start calling them before it joins the cluster (which can happen in `starting` method too)
- under "not normal" circumstances, eg. when pod crashed and restarted, memberlist could receive gossip messages before joining the cluster -- and call these delegate methods before initialization in `starting` has finished. This time window is very tiny, so we wouldn't drop many messages this way.

[the remark this answers is not in the document]
Comment (ortuman) [reply kv/memberlist/memberlist_client.go:237]: Thanks for the clarification Peter. 👍

**Review by pracucci — APPROVED** [review]
Thanks Peter, LGTM! I also think discarding the messages before the delegate is ready is not an issue.

**Pull request merged**